From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Janmm14 <gitconfig1@janmm14.de>
Date: Fri, 19 Nov 2021 08:23:47 +0200
Subject: [PATCH] Add time measurement per event listener method

Results in more useful warnings as listener class is mentioned

diff --git a/event/src/main/java/net/md_5/bungee/event/EventBus.java b/event/src/main/java/net/md_5/bungee/event/EventBus.java
index ac6b9758557e200127a3c03e0d3ef89912963fad..cc0d5e5b9470bdaee8c7ff76c283ec33ecddccf3 100644
--- a/event/src/main/java/net/md_5/bungee/event/EventBus.java
+++ b/event/src/main/java/net/md_5/bungee/event/EventBus.java
@@ -43,6 +43,7 @@ public class EventBus
         {
             for ( EventHandlerMethod method : handlers )
             {
+                long start = System.currentTimeMillis(); // IvanCord
                 try
                 {
                     method.invoke( event );
@@ -58,6 +59,17 @@ public class EventBus
                     logger.log( Level.WARNING, msg, ex.getCause() );
                     if( exceptionHandler != null ) exceptionHandler.handleEventException( msg, event, method, ex ); //Waterfall - call passed exception handler
                 }
+
+                // IvanCord start
+                long elapsed = System.currentTimeMillis() - start;
+                if ( elapsed > 200 )
+                {
+                    logger.log( Level.WARNING, "Plugin listener {0} took {1}ms to process event {2}!", new Object[]
+                    {
+                            method.getListener().getClass().getName(), elapsed, event
+                    } );
+                }
+                // IvanCord end
             }
         }
     }
